File: app/src/main/java/com/example/tanjinc/myapplication/fragment/gridpage/VideoAdapter.java
package com.example.tanjinc.myapplication.fragment.gridpage;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import com.example.tanjinc.myapplication.BaseConstant;
import com.example.tanjinc.myapplication.R;
import com.example.tanjinc.myapplication.bean.ABean;
import com.example.tanjinc.myapplication.bean.BBean;
import com.example.tanjinc.myapplication.bean.BaseBean;
import com.example.tanjinc.myapplication.bean.SkinBean;

public class VideoAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<BaseBean> mDataArray = new ArrayList<>();

    public VideoAdapter() {
    }

    public void setData(List<? extends BaseBean> items) {
        mDataArray.clear();
        mDataArray.addAll(items);
        notifyDataSetChanged();
    }

    public void addFootItems(List<? extends BaseBean> items) {
        mDataArray.addAll(items);
        notifyDataSetChanged();
    }

    public void addHeaderItems(List<? extends BaseBean>  items) {
        mDataArray.addAll(0, items);
        notifyDataSetChanged();
    }


    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        switch (viewType) {
            case BaseConstant.TYPE_A:
                View viewA = layoutInflater.inflate(R.layout.list_item_a, parent, false);
                return new AViewHolder(viewA);
            case BaseConstant.TYPE_B:
                View viewB = layoutInflater.inflate(R.layout.list_item_b, parent, false);
                return new BViewHolder(viewB);
            case BaseConstant.TYPE_VIDEO:
                View view = layoutInflater.inflate(R.layout.list_item_video, parent, false);
                return new VideoViewHolder(view);
        }

        return null;
    }

    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder viewHolder, int position) {
        if (viewHolder == null) {
            return;
        }
        switch (getItemViewType(position)) {
            case BaseConstant.TYPE_A:
                bindTypeAViewHolder(viewHolder, position);
                return;
            case BaseConstant.TYPE_B:
                bindTypeBViewHolder(viewHolder, position);
                break;
            case BaseConstant.TYPE_VIDEO:
                bindTypeVideoViewHolder(viewHolder, position);
        }
    }

    private void bindTypeAViewHolder(final RecyclerView.ViewHolder viewHolder, int position) {
        if(!(viewHolder instanceof AViewHolder)) {
            return;
        }
        AViewHolder aViewHolder = (AViewHolder) viewHolder;
        ABean aBean = (ABean) mDataArray.get(position);
        aViewHolder.mItemName.setText(aBean.getName());
    }

    private void bindTypeBViewHolder(final RecyclerView.ViewHolder viewHolder, int position) {
        if (!(viewHolder instanceof BViewHolder)) {
            return;
        }
        BViewHolder bViewHolder = (BViewHolder) viewHolder;
        BBean bBean = (BBean) mDataArray.get(position);
        bViewHolder.itemNameTv.setText(bBean.getName());
    }

    private void bindTypeVideoViewHolder(final RecyclerView.ViewHolder viewHolder, int position) {
        if (!(viewHolder instanceof VideoViewHolder)) {
            return;
        }
        VideoViewHolder vViewHolder = (VideoViewHolder) viewHolder;
        SkinBean vBean = (SkinBean) mDataArray.get(position);
        vViewHolder.itemNameTv.setText(vBean.getName());
    }

    @Override
    public int getItemCount() {
        return mDataArray.size();
    }


    @Override
    public int getItemViewType(int position) {
        return mDataArray.get(position).getType();
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
        if (layoutManager instanceof GridLayoutManager) {
            GridLayoutManager gridLayoutManager = (GridLayoutManager) layoutManager;
            gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    int type = getItemViewType(position);
                    switch (type) {
                        case BaseConstant.TYPE_A:
                            return 1;
                        case BaseConstant.TYPE_B:
                            return 1;
                        case BaseConstant.TYPE_VIDEO:
                            return 2;
                    }
                    return 3;
                }
            });
        }
        recyclerView.addItemDecoration(new GridItemDecoration());
    }

    private class GridItemDecoration extends RecyclerView.ItemDecoration {

        private Drawable mDivider;

        public GridItemDecoration() {
        }

        @Override
        public void onDraw(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
            final GridLayoutManager layoutManager = (GridLayoutManager) parent.getLayoutManager();
            final GridLayoutManager.SpanSizeLookup lookup = layoutManager.getSpanSizeLookup();


            if (mDivider == null || layoutManager.getChildCount() == 0) {
                return;
            }

            int spanCount = layoutManager.getSpanCount();
            if (layoutManager.getOrientation() == GridLayoutManager.VERTICAL) {

            }
            super.onDraw(c, parent, state);
        }

        @Override
        public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
            super.getItemOffsets(outRect, view, parent, state);
            outRect.bottom = 8;
            outRect.left = 10;
            outRect.right = 10;
        }
    }
}

File: app/src/main/java/com/example/tanjinc/myapplication/fragment/gridpage/VideoAdapter.kt
package com.example.tanjinc.myapplication.fragment.gridpage

import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList

import com.example.tanjinc.myapplication.BaseConstant
import com.example.tanjinc.myapplication.R
import com.example.tanjinc.myapplication.bean.ABean
import com.example.tanjinc.myapplication.bean.BBean
import com.example.tanjinc.myapplication.bean.BaseBean
import com.example.tanjinc.myapplication.bean.SkinBean

class VideoAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val mDataArray = ArrayList<BaseBean>()

    fun setData(items: List<BaseBean>) {
        mDataArray.clear()
        mDataArray.addAll(items)
        notifyDataSetChanged()
    }

    fun addFootItems(items: List<BaseBean>) {
        mDataArray.addAll(items)
        notifyDataSetChanged()
    }

    fun addHeaderItems(items: List<BaseBean>) {
        mDataArray.addAll(0, items)
        notifyDataSetChanged()
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        when (viewType) {
            BaseConstant.TYPE_A -> {
                val viewA = layoutInflater.inflate(R.layout.list_item_a, parent, false)
                return AViewHolder(viewA)
            }
            BaseConstant.TYPE_B -> {
                val viewB = layoutInflater.inflate(R.layout.list_item_b, parent, false)
                return BViewHolder(viewB)
            }
            BaseConstant.TYPE_VIDEO -> {
                val view = layoutInflater.inflate(R.layout.list_item_video, parent, false)
                return VideoViewHolder(view)
            }
            else -> {
                val view = layoutInflater.inflate(R.layout.list_item_video, parent, false)
                return VideoViewHolder(view)
            }
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (viewHolder == null) {
            return
        }
        when (getItemViewType(position)) {
            BaseConstant.TYPE_A -> {
                bindTypeAViewHolder(viewHolder, position)
                return
            }
            BaseConstant.TYPE_B -> bindTypeBViewHolder(viewHolder, position)
            BaseConstant.TYPE_VIDEO -> bindTypeVideoViewHolder(viewHolder, position)
        }
    }

    private fun bindTypeAViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (viewHolder !is AViewHolder) {
            return
        }
        val (name) = mDataArray[position] as ABean
        viewHolder.mItemName.text = name
    }

    private fun bindTypeBViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (viewHolder !is BViewHolder) {
            return
        }
        val bBean = mDataArray[position] as BBean
        viewHolder.itemNameTv.text = bBean.name
    }

    private fun bindTypeVideoViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (viewHolder !is VideoViewHolder) {
            return
        }
        val vBean = mDataArray[position] as SkinBean
        viewHolder.itemNameTv.text = vBean.name
    }

    override fun getItemCount(): Int {
        return mDataArray.size
    }


    override fun getItemViewType(position: Int): Int {
        return mDataArray[position].getType()
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        val layoutManager = recyclerView.layoutManager
        if (layoutManager is GridLayoutManager) {
            val gridLayoutManager = layoutManager as GridLayoutManager?
            gridLayoutManager!!.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    val type = getItemViewType(position)
                    when (type) {
                        BaseConstant.TYPE_A -> return 1
                        BaseConstant.TYPE_B -> return 1
                        BaseConstant.TYPE_VIDEO -> return 2
                    }
                    return 3
                }
            }
        }
        recyclerView.addItemDecoration(GridItemDecoration())
    }

    private inner class GridItemDecoration : RecyclerView.ItemDecoration() {

        private val mDivider: Drawable? = null

        override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
            val layoutManager = parent.layoutManager as GridLayoutManager?
            val lookup = layoutManager!!.spanSizeLookup


            if (mDivider == null || layoutManager.childCount == 0) {
                return
            }

            val spanCount = layoutManager.spanCount
            if (layoutManager.orientation == GridLayoutManager.VERTICAL) {

            }
            super.onDraw(c, parent, state)
        }

        override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
            super.getItemOffsets(outRect, view, parent, state)
            outRect.bottom = 8
            outRect.left = 10
            outRect.right = 10
        }
    }
}
